src/Join.php

```diff
@@ -321,38 +321,42 @@ public function weakJoin($defaults = [])
     /**
      * Creates reference based on a field from the join.
      *
-     * @param Model $model
-     * @param array $defaults
+     * @param string $link
+     * @param array  $defaults
      *
      * @return Reference_One
      */
-    public function hasOne($model, $defaults = [])
+    public function hasOne($link, $defaults = [])
     {
         if (!is_array($defaults)) {
-            $defaults = ['model' => $defaults];
+            $defaults = ['model' => $defaults ?: 'Model_'.$link];
         }
+
         $defaults['join'] = $this;
 
-        return $this->owner->hasOne($model, $defaults);
+        return $this->owner->hasOne($link, $defaults);
     }
 
     /**
      * Creates reference based on the field from the join.
      *
-     * @param Model $model
-     * @param array $defaults
+     * @param string $link
+     * @param array  $defaults
      *
      * @return Reference_Many
      */
-    public function hasMany($model, $defaults = [])
+    public function hasMany($link, $defaults = [])
     {
+        if (!is_array($defaults)) {
+            $defaults = ['model' => $defaults ?: 'Model_'.$link];
+        }
+
         $defaults = array_merge([
             'our_field'   => $this->id_field,
-            'their_field' => $this->table.'_'.$this->id_field,
+            'their_field' => $this->owner->table.'_'.$this->id_field,
```

Comment:
how was this working before? what about recursive joins. ?

Reply:
Most likely it was not working before, because $this (join) ->table property didn't exist, so ... it should throw exception anyway.
Can you add test case for recursive joins? What do you mean? Something like `$this->join($this->table, 'parent_id')` ?